File: Mobile/lib/models/receipt_book.dart
class ReceiptBook {
  final String? bookID;
  final String number;         // Required
  final String type;           // Required
  final String status;         // Required
  final String qrCode;         // Required
  final String? currentHolderID;
  final String? agentID;

  ReceiptBook({
    this.bookID,
    required this.number,
    required this.type,
    required this.status,
    required this.qrCode,
    this.currentHolderID,
    this.agentID,
  });

  factory ReceiptBook.fromJson(Map<String, dynamic> json) {
    return ReceiptBook(
      bookID: json['bookID'] as String?,
      number: json['number'] as String,
      type: json['type'] as String,
      status: json['status'] as String,
      qrCode: json['qrCode'] as String,
      currentHolderID: json['currentHolderID'] as String?,
      agentID: json['agentID'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'bookID': bookID,
      'number': number,
      'type': type,
      'status': status,
      'qrCode': qrCode,
      'currentHolderID': currentHolderID,
      'agentID': agentID,
    };
  }
}

File: -BackUp/keycloak/Back/models/receipt/receiptBook.js
// models/receipt/receiptBook.js
const { nanoid } = require('nanoid');

module.exports = (sequelize, DataTypes) => {
    return sequelize.define('ReceiptBook', {
        bookID: {
            type: DataTypes.STRING,
            primaryKey: true,
            defaultValue: () => `book_${nanoid()}`,
        },
        number: {
            type: DataTypes.STRING,
            allowNull: false,
            unique: true,
        },
        type: {
            type: DataTypes.STRING,
            allowNull: false,
        },
        status: {
            type: DataTypes.ENUM(
                'In Stock',              // Purchase team
                'Sent to Supplier',      // Purchase team
                'Collect from Supplier', // Purchase team
                'With Regional Manager', // Regional manager(s)
                'With Supervisor',       // Supervisor(s)
                'Assigned to Agent',     // Agent
                'Stub Collected',        // Supervisor after collecting stub
                'With Stock Manager',    // Stock manager
                'Archived'               // Final state
            ),
            allowNull: false,
            defaultValue: 'In Stock',
        },
        qrCode: {
            type: DataTypes.BLOB,
            allowNull: false,
        },
        agentID: {
            type: DataTypes.STRING,
            allowNull: true,
            references: { model: 'Agents', key: 'agentID' },
        },
        currentHolderID: {
            type: DataTypes.STRING,
            allowNull: true, // Null when with agent or in stock initially
            references: { model: 'Users', key: 'userID' },
        },
    });
};